app.js

```diff
@@ -1,12 +1,154 @@
 (function() {
 
+  var MACROS_URI = '/api/v2/macros.json';
+
   return {
     events: {
-      'app.activated':'doSomething'
+      'app.activated':                      'initialize',
+      'click .search.btn':                  'startSearch',
+      'requestMacros.done':                 'filterResults',
+      'click .stop.btn':                    'stopSearch'
+    },
+
+    requests: {
+      requestMacros: function(url) {
+        return {
+          url: url || MACROS_URI,
+          type: 'GET',
+          dataType: 'json'
+        };
+      }
+    },
+
+    initialize: function() {
+      this.switchTo('search');
+      this.stopped = true;
+    },
+
+    startSearch: function() {
+      if ( this.$('.check:checked').length < 1 ) {
+        services.notify("Please check at least one condition's checkbox.", 'alert');
+      } else {
+        this.stopped = false;
+        this.$('.stop.btn').show();
+        this.$('.count').text('');
+        this.$('.results ul').empty();
+        this.$('.search.btn').prop('disabled', true);
+        this.$('.query').prop('disabled', true);
+        this.$('.search.btn').prop('value', 'Searching...');
+        this.ajax('requestMacros');
+      }
+      return false;
+    },
+
+    stopSearch: function() {
+      this.stopped = true;
     },
 
-    doSomething: function() {
+    finishSearch: function() {
+      this.$('.search.btn').prop('disabled', false);
+      this.$('.query').prop('disabled', false);
+      this.$('.search.btn').prop('value', 'Search');
+      this.$('.stop.btn').hide();
+      this.stopped = true;
+    },
+
+    filterResults: function(data) {
+      console.log(data);
+      var self = this;
+      var results = data.macros;
+
+      if ( this.$('.check.tag').is(':checked') ) {
+        results = this.filterTagResults(results);
+      }
+      if ( this.$('.check.comment').is(':checked') ) {
+        results = this.filterCommentResults(results);
+      }
+
+      this.displayResults(results);
+
+      // Get additional pages of api request results
+      if (data.next_page && !this.stopped){
+        this.ajax('requestMacros', data.next_page);
+      } else {
+        this.finishSearch();
+      }
+    },
+
+    filterTagResults: function(macros) {
+      var self = this;
+      var results = [];
+      var query = this.$('.query.tag').val();
+
+      macros.forEach( function(macro) {
+        var tags = self.getValues(macro);
+        if ( tags.indexOf(query) > -1 ) {
+          results.push(macro);
+        }
+      });
+
+      return results;
+    },
+
+    filterCommentResults: function(macros) {
+      var self = this;
```

Comment:
Just found something; thought you might be interested: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind

It looks like I can use `bind` to do it instead of using `self = this`:

```
macros.forEach( function(macro) {
  var tags = this.getValues(macro);
    if ( tags.indexOf(query) > -1 ) {
      results.push(macro);
    }
}.bind(this) );
```

I actually came across it in the source code for that Advanced Search app from ZD labs:
https://github.com/zendesklabs/advanced_search/blob/master/app.js#L348